### PR TITLE
[MNG-8146] Drop commons-lang

### PR DESCRIPTION
**Pull request by cstamas**
For start, keeping almost same LOC but the distro zip lost almost 1 MB (was 10MB now is 9MB). Second, am really unsure why it was introduced in the first place, as it merely caused confusion, over StringUtils for example.

---

https://issues.apache.org/jira/browse/MNG-8146

### CODE REVIEW

**Review by michael-o — CHANGES_REQUESTED** [review]
Several changes are *not* OK. There is clearly a definition that for a null value an NPE is expected, not IAE for past 10+ years. This violates the expectations from Effective Java. @elharo

**Comment by michael-o** [comment]
> > Several changes are _not_ OK. There is clearly a definition that for a null value an NPE is expected, not IAE for past 10+ years. This violates the expectations from Effective Java. @elharo
> 
> This is really nitpicking. Can you show me _where we make use of that important distinction_?

Yes, through out our entire codebase.

File: pom.xml [hunk 1]
```diff
@@ -442,16 +441,6 @@ under the License.
...
         <artifactId>commons-cli</artifactId>
         <version>${commonsCliVersion}</version>
-        <exclusions>
-          <exclusion>
-            <groupId>commons-lang</groupId>
-            <artifactId>commons-lang</artifactId>
-          </exclusion>
-          <exclusion>
-            <groupId>commons-logging</groupId>
-            <artifactId>commons-logging</artifactId>
```
Comment (michael-o) [inline pom.xml:452]: Why did you remove the commons-logging exclusion? This is unrelated.

**Comment by michael-o** [comment]
I agree that for those few cases Commons Lang is overkill.

File: pom.xml (hunk 1 above)
Comment (cstamas) [reply pom.xml:452]: Because commons-cli does not depend on any of these, this seems legacy that was never removed? Commons cli deps are only these:
```
  <dependencies>
    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>junit-jupiter-api</artifactId>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>junit-jupiter-engine</artifactId>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>org.junit.jupiter</groupId>
      <artifactId>junit-jupiter-params</artifactId>
      <scope>test</scope>
    </dependency>
    <dependency>
      <groupId>commons-io</groupId>
      <artifactId>commons-io</artifactId>
      <version>2.16.1</version>
      <scope>test</scope>
    </dependency>
  </dependencies>
```

Comment (michael-o) [reply pom.xml:452]: Commons Logging has been removed as a dep?

**Comment by cstamas** [comment]
> Yes, through out our entire codebase.

Don't divert: I asked you to show me where _changed_ spots in this PR benefit anything from NPE (as vs IAEx). 

But nevertheless, with copy-paste (of message) introduced Objects.requireNonNull use instead to do throw NPE "by the book".

**Review by rmannibucau — APPROVED** [review]

**Comment by elharo** [comment]
I'll have to dig into the individual changes. However overall I no longer trust commons-lang and definitely favor removing our dependence on it if possible.

File: maven-artifact/src/main/java/org/apache/maven/artifact/versioning/DefaultArtifactVersion.java [hunk 2]
```diff
@@ -176,6 +174,18 @@ private static Integer getNextIntegerToken(StringTokenizer tok) {
         return tryParseInt(s);
     }
 
+    private static boolean isDigits(String str) {
+        if (str == null || str.trim().isEmpty()) {
+            return false;
+        }
+        for (int i = 0; i < str.length(); i++) {
+            if (!Character.isDigit(str.charAt(i))) {
```
Comment (elharo) [inline maven-artifact/src/main/java/org/apache/maven/artifact/versioning/DefaultArtifactVersion.java:182]: This is going to allow digits from other alphabets beyond the usual 0-9. I don't think we want that, do we?

**Comment by cstamas** [comment]
For ref, the Maven4 PR https://github.com/apache/maven/pull/1080

File: maven-core/src/main/java/org/apache/maven/rtinfo/internal/DefaultRuntimeInformation.java [hunk 3]
```diff
@@ -80,10 +79,13 @@ public String getMavenVersion() {
     }
 
     public boolean isMavenVersion(String versionRange) {
+        if (Objects.requireNonNull(versionRange, "versionRange can neither be null, empty nor blank")
```
Comment (elharo) [inline maven-core/src/main/java/org/apache/maven/rtinfo/internal/DefaultRuntimeInformation.java:82]: nit: comma after empty

File: maven-core/src/main/java/org/apache/maven/configuration/DefaultBeanConfigurationRequest.java [hunk 4]
```diff
@@ -104,8 +105,16 @@ public DefaultBeanConfigurationRequest setConfiguration(
     }
 
     private Plugin findPlugin(Model model, String groupId, String artifactId) {
-        Validate.notBlank(groupId, "groupId can neither be null, empty nor blank");
-        Validate.notBlank(artifactId, "artifactId can neither be null, empty nor blank");
+        if (Objects.requireNonNull(groupId, "groupId can neither be null, empty nor blank")
```
Comment (elharo) [inline maven-core/src/main/java/org/apache/maven/configuration/DefaultBeanConfigurationRequest.java:108]: nit: comma after empty

File: maven-embedder/src/test/java/org/apache/maven/cli/MavenCliTest.java [hunk 5]
```diff
@@ -89,7 +89,10 @@ public void testCalculateDegreeOfConcurrency() {
         int cpus = Runtime.getRuntime().availableProcessors();
         assertEquals((int) (cpus * 2.2), cli.calculateDegreeOfConcurrency("2.2C"));
         assertEquals(1, cli.calculateDegreeOfConcurrency("0.0001C"));
-        assertThrows(IllegalArgumentException.class, new ConcurrencyCalculator("2.C"));
+        // Note: this assertion below makes no sense, first Float.parseFloat("2.") DOES parse the string in 2.0.
```
Comment (elharo) [inline maven-embedder/src/test/java/org/apache/maven/cli/MavenCliTest.java:92]: OK, just delete this then

**Comment by elharo** [comment]
I concur that NPE is strongly preferred over IAE for null values. I didn't happen to find any places where this was done wrong, but I wasn't reading that closely.

**Comment by cstamas** [comment]
I agree with this "general" rule to prefer NPE over IAE, but disagree with unconditional sticking to this rule. One has to be pragmatic a bit, and allow deviation. My point was, that in this PR, the spots where it was changed did and does not benefit _anything_ from distinguishing NPE from IAE, while if you look at change that made PR "by the book" (to distinguish NPE and IAE) worsened the quality of code, as it caused string/message duplication all over the place.

So, IMHO rigidly sticking to something is just as bad as never adhering to any best practice.

**Review by michael-o — APPROVED** [review]
Very good now. What about transitive commons lang?

**Review by slachiewicz — APPROVED** [review]

**Review by slawekjaranowski — APPROVED** [review]

File: maven-artifact/src/main/java/org/apache/maven/artifact/versioning/DefaultArtifactVersion.java [hunk 6]
```diff
@@ -179,7 +179,8 @@ private static boolean isDigits(String str) {
             return false;
         }
         for (int i = 0; i < str.length(); i++) {
-            if (!Character.isDigit(str.charAt(i))) {
+            char c = str.charAt(i);
+            if (!(c >= '0' && c <= '9')) {
```
Comment (michael-o) [inline maven-artifact/src/main/java/org/apache/maven/artifact/versioning/DefaultArtifactVersion.java:183]: Good!

File: maven-artifact/src/test/java/org/apache/maven/artifact/versioning/DefaultArtifactVersionTest.java [hunk 7]
```diff
@@ -80,6 +80,15 @@ public void testVersionParsing() {
...
     }
 
+    public void testVersionParsingNot09() {
+        String ver = "१.२.३";
+        assertTrue(Character.isDigit(ver.charAt(0)));
+        assertTrue(Character.isDigit(ver.charAt(2)));
+        assertTrue(Character.isDigit(ver.charAt(4)));
+        ArtifactVersion version = newArtifactVersion(ver);
+        assertEquals(ver, version.getQualifier());
+    }
```
Comment (michael-o) [inline maven-artifact/src/test/java/org/apache/maven/artifact/versioning/DefaultArtifactVersionTest.java:90]: Hold on, shouldn't this fail?

Comment (cstamas) [reply maven-artifact/src/test/java/org/apache/maven/artifact/versioning/DefaultArtifactVersionTest.java:90]: No, as these are not numbers anymore, they are qualifiers (otherwise it would go into major, minor...)

Comment (michael-o) [reply maven-artifact/src/test/java/org/apache/maven/artifact/versioning/DefaultArtifactVersionTest.java:90]: man, I want `U+1F4A9` as qualifier!

Comment (cstamas) [reply maven-artifact/src/test/java/org/apache/maven/artifact/versioning/DefaultArtifactVersionTest.java:90]: but we must carry thru same change on Resolver as well...
